Mitigate JSON writing performance regression

timwoj opened the pull request:
Fixes #595 

Requires https://github.com/zeek/zeek-3rdparty/pull/3 to be merged

The times below are using the test script and command line from the GHI linked above.

On `master`:

```
real	0m21.563s
user	0m25.880s
sys	0m5.525s
```

With changes on this branch:

```
real	0m13.034s
user	0m17.429s
sys	0m5.499s
```

Bro 2.6.4:
```
real	0m12.878s
user	0m16.419s
sys	0m6.537s
```

timwoj commented:
Added bro 2.6.4 metrics.

jsiwek commented on src/Val.cc:
```diff
@@ -646,9 +646,8 @@ static ZeekJson BuildJSON(Val* val, bool only_loggable=false, RE_Matcher* re=new
-			j = ZeekJson::object();
 			auto* oval = val->AsOpaqueVal();
-			j["opaque_type"] = OpaqueMgr::mgr()->TypeID(oval);
+			j = {"opaque_type", OpaqueMgr::mgr()->TypeID(oval)};
```
I expected a map-like initializer to look like `{ {key, value} }`, not `{key, value}`, so thought this change might actually be making an array now instead of an object ?  This example seems to confirm:

```
ZeekJson j;
j = { {"one", "two"} };
printf("%s\n", j.dump().data());
j = {"one", "two"} ;
printf("%s\n", j.dump().data());
```

Outputs:

```
{"one":"two"}
["one","two"]
```

timwoj replied:
Yep, you're right. I misread the docs. I think there's also a missing unit test there since none of them caught that.

jsiwek commented on src/threading/formatters/JSON.cc:
```diff
@@ -182,11 +182,7 @@ ZeekJson JSON::BuildJSON(Value* val, const string& name) const
-		j2[name] = j;
-		return j2;
-		}
+		return { name, j };
```
This makes an array, not an object now?

jsiwek commented:
Also, the license in `COPYING.3rdparty` should be updated (remove fifo_map, add tsl).

timwoj commented:
Fixed the above comments. I added a unit test for the change in Val.cc. The block in JSON.cc that changed is never actually called, but it's correct now. I made a note to go back and look at whether that block can just be removed.

rsmmr commented:
Can we get this evaluated in one of the environments that had performance problems with the 3.0.0 release to see if it addresses the problem sufficiently? Note we should also include any other JSON changes (thinking of #611), so maybe just start merging it all into the 3.0 branch?

sethhall commented:
I'll start pushing on that.

> On Oct 1, 2019, at 5:13 AM, Robin Sommer <notifications@github.com> wrote:
> 
> ﻿
> Can we get this evaluated in one of the environments that had performance problems with the 3.0.0 release to see if it addresses the problem sufficiently? Note we should also include any other JSON changes (thinking of #611), so maybe just start merging it all into the 3.0 branch?
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

jsiwek commented:
I'll work on getting this merged along with other things into the `release/3.0` branch to make it easier for people to test (and after, I'll also ping Keith directly to help test).
